org.eclipse.xtext/src/org/eclipse/xtext/generator/URIBasedFileSystemAccess.java

```diff
@@ -0,0 +1,222 @@
+/**
+ * Copyright (c) 2015, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.generator;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.UnsupportedEncodingException;
+import java.util.Collections;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.resource.URIConverter;
+import org.eclipse.xtext.generator.trace.ITraceRegionProvider;
+import org.eclipse.xtext.generator.trace.TraceFileNameProvider;
+import org.eclipse.xtext.generator.trace.TraceNotFoundException;
+import org.eclipse.xtext.generator.trace.TraceRegionSerializer;
+import org.eclipse.xtext.parser.IEncodingProvider;
+import org.eclipse.xtext.util.RuntimeIOException;
+
+import com.google.common.io.ByteStreams;
+import com.google.common.io.CharStreams;
+
+/**
+ * A file system access implementation that is based on EMF URIs and URIConverter
+ * 
+ * @since 2.9
+ */
+public class URIBasedFileSystemAccess extends AbstractFileSystemAccess2 {
+	public interface BeforeDelete {
+		/**
+		 * @return <code>true</code> if the file can be deleted, false otherwise
+		 */
+		boolean beforeDelete(URI changed);
+	}
+
+	public interface BeforeWrite {
+		InputStream beforeWrite(URI changed, String outputCfgName, InputStream in);
```

Comment:
It would be really useful if these would declare `throws IOException`

Reply:
Not sure here about API consequences ... and what about BeforeRead and BeforeDelete!?

Reply:
I tried to include `beforeRead` and `beforeDelete` in `these`.
Declaring an exception on the signature would be binary compatible, but source incompatible. However, since some the callbacks will be implemented with Xtend, it wouldn't matter (sneakyThrow ...). @cdietrich What's your take?

Reply:
is there anybody out there using this at all?

Reply:
maybe you can have a look at these